wc: streaming --files0-from and other improvements

jeddenlea opened the pull request:
My original focus was on `--files0-from`, which should be handled as a stream if the file specified (or stdin) is not a regular file. I've accomplished this by making a new `Inputs` type to represent the desired list of files to process. Its `try_iter` method does most of the heavy lifting to figure out how to actually run.

Secondarily, I have attempted to reduce the number of String allocations that occur while printing the results. Now, unless a file name needs escaping, or an error occurs, no additional allocations should be necessary to print results. An `Input` now borrows file names from the command line, unless they're read from `--files0-from`. `print_stats` was the biggest abuser, allocating a `String` for each column before before being `join`'d into a `String` for the whole line.  The `TitledWordCount` type is not necessary at all, which was another source of `String` allocation.

Finally, I've made some effort to make more cases match GNU wc's output. Errors encountered processing `--files0-from` as well as any encountered processing files listed on the command line will be escaped more consistently like GNU wc.  File names printed to the right of their stats will be less aggressively quoted, to match GNU wc, now only if they are not UTF-8 or if they contain a newline.

cakebaker commented:
Can you please run `cargo clippy`? Thanks!

jeddenlea commented:
Yes. Apologies, I thought I had!

On Fri, Mar 31, 2023, 01:59 Daniel Hofstetter ***@***.***>
wrote:

> Can you please run cargo clippy? Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uutils/coreutils/pull/4696#issuecomment-1491571856>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOVRSXXI6M6D5LVCYTOQ5LW62MGVANCNFSM6AAAAAAWOJ2MOM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jeddenlea commented:
Yes. Apologies, I thought I had!

sylvestre commented:
could you please split the PR into smaller commits ? This is a big patch ...

jeddenlea commented:
I've broken up the change into 4 different commits.  The lines between some concerns were a little blurry, as I ended up touching a lot of stuff, but I've tried my best to keep them each cleanly readable on their own.

I'm still working on getting the GNU coreutils tests running completely, but I've at least got the one complaining to run.  And, it turns out `wc` will read the entirety of a `--files0-from` file before processing so long as it is no more than `10MiB`, so I need to do a little refactoring on the last commit...

sylvestre commented:
please remove the "draft" when ready to be reviewed :)

jeddenlea commented:
Alright, I think I'm ready for some feedback to iterate on here.  On one point in particular, I didn't realize that when I introduced use of AsRawFd/FromRawFd that the import paths were new as of 1.66.  Also, the handy `ilog10` method for determining how many digits an int will need for printing was only added in 1.67!  I see the current minimum is 1.64, which is pretty recent compared to some other repos' minimum version.  What's your policy on updating the minimum version of Rust?

tertsdiepraam commented:
> What's your policy on updating the minimum version of Rust?

We do it only after cutting a release and check how important the new features are. Essentially we bump it by need. But always in a separate PR.

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -621,63 +621,52 @@ fn compute_number_width(inputs: &[Input], settings: &Settings) -> usize {
 }
 
+/// shorthand for `show!(USimpleError::new(...))`, works just like `format!`.
+macro_rules! record_error {
```
Not convinced of the usefulness of this over just using the code. The verboseness of that code should actually incentivize doing the error handling with a custom error type :)

jeddenlea replied:
Ahhh, yeah, best to avoid making the `USimpleError` at all if we can...  Was that your major objection, or do would you not be okay with,
```rust
macro_rules! record_error {
    ($fmt:literal $(, $arg:expr)*) => {
        eprintln!($fmt $(, $arg)*);
        uucore::errors::set_exit_code(1);
    }
}
```
?

tertsdiepraam replied:
I think we should use the code without the custom macro here. However, we can talk in another issue or PR about adding this to uucore instead. At the same time, we can then see whether we can give `show!` a better name. Does that sound good?

Also I made a mistake yesterday, it should be `show_error!` (again a pretty confusing name) not `eprintln!`, because we want to add the name of the util at the start of the line.

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -688,29 +677,35 @@ fn wc(inputs: &[Input], settings: &Settings) -> UResult<()> {
-        columns.push(title.clone());
+    let mut space = "";
+
+    macro_rules! print_column {
```
Here again the macro is not really pulling its weight in my opinion. It could also be expressed as a closure:
```rust
let mut space = "";

let mut print_column = |show: bool, result: usize| -> io::Result<()> {
    if show {
        write!(stdout, "{space}{:width$}", result, width = number_width)?;
        space = " ";
    }
    Ok(())
};

print_column(settings.show_lines, result.lines)?;
print_column(settings.show_words, result.words)?;
print_column(settings.show_chars, result.chars)?;
print_column(settings.show_bytes, result.bytes)?;
print_column(settings.show_max_line_length, result.max_line_length)?;
```

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -538,9 +553,9 @@ enum CountResult {
+    Interrupted(WordCount, UIoError),
     /// Didn't even manage to open.
-    Failure(io::Error),
+    Failure(UIoError),
```
This seems weird, because then we have the context of the `UIoError` wrapped in the context provided by the enum? Conversion from `io::Error` to `UError` to `WcError` to `UError` seems like a weird way to do things.

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -592,8 +607,10 @@ fn word_count_from_input(input: &Input, settings: &Settings) -> CountResult {
+        || settings.number_enabled() == 1
+            && (inputs.len() == 1
+                || settings.files0_from_path.as_ref().map(|p| p.as_os_str())
+                    == Some(OsStr::new(STDIN_REPR)))
```
Oof this formatting is a bit rough. Is that really what rustfmt wants? Maybe you could break up into a few smaller booleans first?

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -210,33 +223,33 @@ impl TotalWhen {
 }
 
-#[derive(Debug)]
+#[derive(Debug, Error)]
```
`thiserror` is not doing a whole lot here except here in my opinion, but it's a big deal, because we already use `thiserror` elsewhere.

tertsdiepraam replied:
> When thiserror fits, I think it's a pretty nice method of creating errors.

Yeah, I kinda agree. We should probably get some project-wide standard on this. At the moment it's mostly that most utils use no library and then there's some `thiserror` and even `quick-error` in some utils, which I think I a bit annoying. I think it's easiest to converge on not using a library, because we rarely need the fancy features of `thiserror`. Here it's also just implementing `Display`.

With errors you usually end up cloning because they tend to escape the scope where they were created.

Btw, I made an edit to my original comment, because I made a very significant typo:
> but it's a big deal,

should have been

> but it's not a big deal,

Sorry if that caused some confusion :)

jeddenlea replied:
Heh, I think I only noticed that the third time I read it :)

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -41,17 +42,17 @@ use crate::{
     show_words: bool,
     show_max_line_length: bool,
-    files0_from_path: Option<PathBuf>,
+    files0_from: Option<Input<'a>>,
```
If you make things more complicated for performance reasons (which I assume the lifetimes are for), you do kind of need to show that performance improves. I bet it doesn't really matter, because it happens just once and this is not a hot path. The same applies to some of the `Cow` strings. If you just get rid of cloning and do not introduce complexity it's always fine of course.

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -558,84 +654,148 @@ enum CountResult {
-    max(minimum_width, total.to_string().len())
+#[derive(Debug, Error)]
+#[error("{0}")]
+enum InputError {
```
`UError` already includes `io::Error`.

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -648,22 +808,27 @@ macro_rules! record_error {
-        }
-        let word_count = match word_count_from_input(input, settings) {
+    for maybe_input in inputs.try_iter(settings)? {
+        num_inputs = num_inputs.saturating_add(1);
```
This seems a bit overly defensive. If `usize` is gonna overflow, we have other more important problems :)

tertsdiepraam reviewed (commented):
Pretty cool stuff! I would suggest that you keep patches much smaller in the future though.

jeddenlea reviewed (commented):
Thanks for the feedback!  I've done my best to address your concerns within each of the 4 commits, and I've rebased them all atop main.

sylvestre commented:
See this failure:
```

     Running `sccache rustc --crate-name uu_relpath --edition=2021 src/uu/relpath/src/relpath.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C metadata=9e8dee9d6aa6280e -C extra-filename=-9e8dee9d6aa6280e --out-dir /home/runner/work/coreutils/coreutils/target/debug/deps -L dependency=/home/runner/work/coreutils/coreutils/target/debug/deps --extern clap=/home/runner/work/coreutils/coreutils/target/debug/deps/libclap-e00c16ba6075349e.rmeta --extern uucore=/home/runner/work/coreutils/coreutils/target/debug/deps/libuucore-1bf9a7e8458cc252.rmeta -Awarnings -L native=/home/runner/work/coreutils/coreutils/target/debug/build/blake3-ca32ffe4cb3ec7ed/out -L native=/home/runner/work/coreutils/coreutils/target/debug/build/blake3-ca32ffe4cb3ec7ed/out`
error[E0599]: no method named `ilog10` found for type `u64` in the current scope
   --> src/uu/wc/src/wc.rs:719:46
    |
719 |                 let total_width = (1 + total.ilog10())
    |                                              ^^^^^^ help: there is an associated function with a similar name: `log`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `uu_wc` due to previous error

Caused by:
```

jeddenlea commented:
Hey Sylvestere.  Yeah, that's what prompted my question about your policy around minimum versions.... If you anticipate moving to 1.67 soon, I'd leave that in place.  Otherwise, we'll need to do something different.  It's unfortunate Rust hasn't stabilized its version check, https://doc.rust-lang.org/beta/unstable-book/language-features/cfg-version.html

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -113,68 +125,216 @@ pub mod options {
-impl From<&OsStr> for Input {
-    fn from(input: &OsStr) -> Self {
-        if input == STDIN_REPR {
+impl From<PathBuf> for Input<'_> {
```
As far as I can tell, this impl is not used right now.

jeddenlea replied:
You are right!  The uglier one below is taking precedence now....

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -113,68 +125,216 @@ pub mod options {
 }
 
-impl Input {
+impl<'a, T: AsRef<Path> + ?Sized> From<&'a T> for Input<'a> {
```
`?Sized` is unnecessary here I think.

jeddenlea replied:
I swear it was at one point, but I guess I cleaned up :)

tertsdiepraam commented on src/uu/wc/src/wc.rs:
```diff
@@ -113,68 +125,216 @@ pub mod options {
+
+    // Creates an iterator which yields values borrowed from the command line arguments.
+    // Returns an error if the file specified in --files0-from cannot be opened.
+    fn try_iter(
+        &'a self,
+        settings: &'a Settings<'a>,
```
The structure of functions here is what gets you into trouble with a lot of references and the `Cow` in `Input`. What if you make `Input` always a reference and then structure it like this:
```rust
fn count_inputs(inputs: Inputs) -> ... {
    match inputs {
        Inputs::Stdin => count_input(Input::Stdin(StdinKind::Implicit)),
        Inputs::Paths(paths) => /* call count_input for each path */,
        Self::Files0From(path) => /* call count_input for each line */
}
```
Then you don't have to pass `Iterators` around which makes lifetimes a lot easier.

jeddenlea replied:
The iterator returned by `try_iter` takes care of so much, though.  It also helps keep the flow of the program consistent with what was happening before.  It used to be that it would only iterate through a `[Input]`, then we couldn't handle streams.  I wanted to be able to handle streams from `--files0-from` in the same manner as a list of command line arguments, and it has also made it easy to consistently handle errors like empty file names.

I should have mentioned already, and I'm embarrassed I still haven't gotten to it yet, but it was my intent to write some more tests to capture the new cases in which this change makes this `wc` act like GNU's.  The formatting of the errors when empty paths are found in either a `--files0-from` or the command line is one "feature" we get with this.

tertsdiepraam commented:
> If you anticipate moving to 1.67 soon

I think not just for this.

> It's unfortunate Rust hasn't stabilized its version check

Honestly, I'm glad 😄 At least for this project, keeping a consistent MSRV would always be the best option. For libraries, it might be different of course.

jeddenlea commented:
> > If you anticipate moving to 1.67 soon
> 
> I think not just for this.

Oh sure, this is not a killer feature worth it on its own.  I'll see about copying it in verbatim if it's short, or writing a facsimile... 

> 
> > It's unfortunate Rust hasn't stabilized its version check
> 
> Honestly, I'm glad 😄 At least for this project, keeping a consistent MSRV would always be the best option. For libraries, it might be different of course.

Those jerks just keep adding useful stuff to the standard library!  This is certainly not the first time I stumbled onto something useful that just happened to be very new and caused this problem.

jeddenlea commented on src/uu/wc/src/wc.rs:
```diff
@@ -521,187 +645,223 @@ enum CountResult {
+        let input = match maybe_input {
+            Ok(input) => input,
+            Err(err) => {
+                record_error!("{err}");
```
Yeah that's slick!  So now not only do we get to punt on the naming of show, but we still get to make each case of it into a single line which was the whole point of `record_error` to begin with.

jeddenlea commented on src/uu/wc/src/wc.rs:
```diff
@@ -521,187 +645,223 @@ enum CountResult {
-                    )
-                ));
+            CountResult::Interrupted(word_count, err) => {
+                record_error!("{}: {}", input.path_display(), err);
```
`path_display` is still necessary for consistent errors, but that's actually shorter.

sylvestre commented:
@jeddenlea on android:
```
 error: pathspec './tests/fixtures/wc/alice' did not match any file(s) known to git
error: pathspec 'in' did not match any file(s) known to git
```
rings a bell ?

jeddenlea commented:
> @jeddenlea on android:
> 
> ```
>  error: pathspec './tests/fixtures/wc/alice' did not match any file(s) known to git
> error: pathspec 'in' did not match any file(s) known to git
> ```
> 
> rings a bell ?

I had made a symlink named `alice in wonderland.txt` that pointed to `alice_in_wonderland.txt`, hoping to emphasize that their contents were the same.  But, that Android script is not friendly to symlinks with whitespace in their name.  So, to keep things simple, I'll just make `alice in wonderland.txt` a file too.

jeddenlea commented on tests/by-util/test_wc.rs:
```diff
@@ -1,6 +1,6 @@
 use crate::common::util::{vec_of_size, TestScenario};
 
-// spell-checker:ignore (flags) lwmcL clmwL ; (path) bogusfile emptyfile manyemptylines moby notrailingnewline onelongemptyline onelongword weirdchars
+// spell-checker:ignore (flags) lwmcL clmwL ; (path) bogusfile emptyfile manyemptylines moby notrailingnewline onelongemptyline onelongword weirdchars nonexistant
```
I spelled it wrong :grimacing:

jeddenlea commented:
The GnuTest failures look like a hiccup with `apt`.  Hopefully the Windows tests were also just bad luck...

sylvestre commented:
nope, I think the test doesn't work on windows. see:

```
  TRY 1 SLOW [>4800.000s] coreutils::tests test_wc::test_files0_progressive_stream
   Canceling due to interrupt: 1 tests still running
       ABORT [4856.604s] coreutils::tests test_wc::test_files0_progressive_stream
     Message [         ] code 0xc000013a: {Application Exit by CTRL+C}
The application terminated as a result of a CTRL+C. (os error 572)
------------
     Summary [5131.687s] 2147 tests run: 2140 passed (1 slow, 4 leaky), 7 failed, 24 skipped
error: test run failed
Error: The operation was canceled.

```

it times out

jeddenlea commented:
> nope, I think the test doesn't work on windows. see:
> 
> ```
>   TRY 1 SLOW [>4800.000s] coreutils::tests test_wc::test_files0_progressive_stream
>    Canceling due to interrupt: 1 tests still running
>        ABORT [4856.604s] coreutils::tests test_wc::test_files0_progressive_stream
>      Message [         ] code 0xc000013a: {Application Exit by CTRL+C}
> The application terminated as a result of a CTRL+C. (os error 572)
> ------------
>      Summary [5131.687s] 2147 tests run: 2140 passed (1 slow, 4 leaky), 7 failed, 24 skipped
> error: test run failed
> Error: The operation was canceled.
> ```
> 
> it times out

:facepalm: I guess I could have looked myself!  Sorry about that.  I think I understand what the failure is, and it shouldn't be too hard to fix in theory.  I'm currently in the middle of getting VirtualBox running so I can actually work on the thing in Windows....

jeddenlea commented:
Woo! Looks like I fixed the Windows stuff.  That was a.... pain.  VirtualBox didn't work as it is wont to do.  I ended up using a free t2.micro on AWS, probably the first time I've used Windows since I almost threw a Windows 7 laptop out of my window because it was so bloody slow and awful!

Anyway, in the interest of preventing any individual commits entering the tree which fail CI on their own, I patched the last two here and rebased them all again.  You can see the diffs here: https://github.com/jeddenlea/coreutils/compare/old_wc..wc

sylvestre commented:
impressive, well done :)
